sw/device/silicon_creator/lib/cert/asn1.h

```diff
@@ -0,0 +1,216 @@
+ * @param buffer Pointer to a user-provided buffer.
+ * @param size Size of the user-provided buffer.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_start(asn1_state_t *new_state, uint8_t *buffer, size_t size);
+
+/**
+ * Finish an ASN1 stream and return the size.
+ *
+ * Note: the state will be cleared out after this call.
+ *
+ * @param state Pointer to the state initialized by asn1_start.
+ * @param[out] out_size Pointer to an integer that will be set to the size of
+ * the stream.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_finish(asn1_state_t *state, size_t *out_size);
+
+/**
+ * Push a byte into the ASN1 buffer.
+ *
+ * @param state Pointer to the initialized by asn1_start.
+ * @param byte Byte to add to the buffer.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_push_byte(asn1_state_t *state, uint8_t byte);
+
+/**
+ * Push bytes into the ASN1 buffer.
+ *
+ * @param state Pointer to the initialized by asn1_start.
+ * @param bytes Pointer to an array of bytes.
+ * @param size Number of bytes in the array.
+ * @return The result of the operation.
+ */
+rom_error_t asn1_push_bytes(asn1_state_t *state, const uint8_t *bytes,
+                            size_t size);
+
+/**
+ * Structure holding the information about an unfinished tag sequence.
+ *
+ * The fields in this structure should be considered
+ * private and not be read or written directly.
+ */
+typedef struct asn1_tag {
+  // Pointer to state.
+  asn1_state_t *state;
+  // Offset of the start of the length octets.
+  size_t len_offset;
+  // How many bytes were allocated for the length octets.
+  size_t len_size;
+} asn1_tag_t;
+
+// ASN1 tag classes.
+typedef enum asn1_tag_class {
+  kAsn1TagClassUniversal = 0 << 6,
+  kAsn1TagClassApplication = 1 << 6,
+  kAsn1TagClassContext = 2 << 6,
+  kAsn1TagClassPrivate = 3 << 6,
+} asn1_tag_class_t;
+
+// ASN1 tag primitive/constructed bit.
+typedef enum asn1_tag_form {
+  kAsn1TagFormPrimitive = 0 << 5,
+  kAsn1TagFormConstructed = 1 << 5,
+} asn1_tag_form_t;
+
+// ASN1 tag number (for universal tags).
+typedef enum asn1_tag_number {
+  kAsn1TagNumberBoolean = 0x01,
+  kAsn1TagNumberInteger = 0x02,
+  kAsn1TagNumberBitString = 0x03,
+  kAsn1TagNumberOctetString = 0x04,
+  kAsn1TagNumberOid = 0x06,
+  kAsn1TagNumberUtf8String = 0x0c,
+  kAsn1TagNumberPrintableString = 0x13,
+  kAsn1TagNumberGeneralizedTime = 0x18,
+  kAsn1TagNumberSequence = 0x30,
+  kAsn1TagNumberSet = 0x31,
+} asn1_tag_number_t;
+
+/**
+ * Start an ASN1 tag.
+ *
+ * @param state Pointer to the state initialized by asn1_start.
+ * @param[out] new_tag Pointer to a user-allocated tag to be initialized.
+ * @param id Identifier byte of the tag (see ASN1_CLASS_*, ASN1_FORM_* and
+ * ASN1_TAG_*).
+ * @return The result of the operation.
+ */
+rom_error_t asn1_start_tag(asn1_state_t *state, asn1_tag_t *new_tag,
+                           uint8_t id);
```

Comment:
I think that x509 does not use any multibyte tag (Although I don't have much experience with it). However, we never know where else we may want to use this library, one example that I worked with is the [EMVCo](https://www.emvco.com/specifications/) specification for payments. 
So maybe it would not hurt changing the `id` type to `uint32_t`?
But I agree that we can always create a new function.

Reply:
Sure I think this is reasonable. I think in this case we could split the 32-bits as follows:
- top 2 bits (31:30) are the class
- bit 29 is constructed 
- bits 28:0 are the tag number
This only requires the update the function and the definiton of `kAsn1*` enums.
Then the function takes care of building the identifier octets. What do you think?

Reply:
That works, or you could go even simpler. As we will only use 1byte tags for now, the function can cast the id to byte. The proper formatting should be checked when/if multibyte tag is properly implemented.
I'll let you decide.